nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,152 @@ defmodule Nx.BinaryBackend do
     from_binary(out, output_data)
   end
 
+  @impl true
+  def qr(
+        {%{shape: {m, k} = shape_q, type: output_type} = q_holder,
+         %{shape: {k, n} = shape_r, type: output_type} = r_holder},
+        %{type: input_type} = tensor,
+        opts
+      ) do
+    mode = opts[:mode]
+
+    r =
+      if mode == :reduced do
+        tensor[[0..(k - 1), 0..(n - 1)]]
+      else
+        tensor
+      end
+      |> to_binary()
+
+    shape_h = {k, k}
+
+    {q_bin, r_bin, _output_type} =
+      for i <- 0..(n - 1), reduce: {nil, r, input_type} do
+        {q, r, type_r} ->
+          # a = r[[i..(k - 1), i]]
+          {a_reverse, num_rows_a, _, _} =
+            match_types [type_r] do
+              for <<match!(x, 0) <- r>>, reduce: {[], 0, 0, 0} do
+                {acc, num_rows_a_prev, row, col} ->
+                  {acc, num_rows_a} =
+                    if row >= i and col == i do
+                      {[read!(x, 0) | acc], num_rows_a_prev + 1}
+                    else
+                      {acc, num_rows_a_prev}
+                    end
+
+                  {row, col} =
+                    if col + 1 == n do
+                      {row + 1, 0}
+                    else
+                      {row, col + 1}
+                    end
+
+                  {acc, num_rows_a, row, col}
+              end
+            end
+
+          h_bin = householder_reflector(a_reverse, num_rows_a, k, output_type)
+
+          # If we haven't allocated Q yet, let Q = H1
+          q =
+            if is_nil(q) do
+              padding_length = (m - k) * k
+
+              zero = number_to_binary(0, output_type)
+
+              h_bin <> String.duplicate(zero, padding_length)
+            else
+              dot_binary(q, shape_q, output_type, h_bin, shape_h, output_type, output_type)
+            end
+
+          r = dot_binary(h_bin, shape_h, output_type, r, shape_r, type_r, output_type)
+
+          {q, r, output_type}
+      end
+
+    q = from_binary(q_holder, q_bin)
+    r = from_binary(r_holder, r_bin)
+
+    case mode do
+      :complete ->
+        {q, r}
+
+      :reduced ->
+        {q, r[[0..(k - 1), 0..(n - 1)]]}
```

Comment:
Is it possible to do the slicing on the binary before building the output tensor? If you look at `Nx.slice` it's just a shape change then a weighted traverse.